src/services/services.ts

```diff
@@ -992,9 +994,13 @@ namespace ts {
             return !isString(entry) ? entry : undefined;
         }
 
-        public getOrCreateEntryByPath(fileName: string, path: Path): HostFileInformation {
-            const info = this.getEntryByPath(path) || this.createEntry(fileName, path);
-            return isString(info) ? undefined! : info; // TODO: GH#18217
+        public getOrCreateEntryByPath(fileName: string, path: Path, isAuxiliaryFile: boolean): HostFileInformation {
+            let info = this.getEntryByPath(path) || this.createEntry(fileName, path, isAuxiliaryFile);

```

Comment:
I wanted to update the program as early as possible so it’s not happening while the user waits for completions. In practice we’re only seeing ~20ms updates so I can try this. I think  we’ll want to do the initial program creation eagerly (maybe triggered from Project) but for future updates I think doing it on demand makes sense.

Reply:
I dont think we should create it eagerly.. Since many times project gets opened only for navigation purposes and creating this would add to the cost.. Instead doing it during completion will impact performance of first completion but later should be ok..  we already have issues with us opening project /creating project when opening the file instead of first request..

Reply:
The navigation case is already handled by checking for open files in the project. There’s a unit test validating that. Creating the project on the first completions request took the response time from ~100ms to ~400ms.